src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,35 +1,64 @@
 package com.es.phoneshop.model.product;
 
-import java.math.BigDecimal;
-import java.util.ArrayList;
-import java.util.Currency;
-import java.util.List;
+import java.util.*;
+import java.util.function.ToIntFunction;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 public class ArrayListProductDao implements ProductDao {
-
     private List<Product> products = new ArrayList<>();
     private long maxId;
+    private Map<SortField, Comparator<Product>> sortComparatorMapping = Map.of(
+            SortField.DESCRIPTION, Comparator.comparing(product -> product.getDescription().toLowerCase()),
+            SortField.PRICE, Comparator.comparing(Product::getPrice)
+    );
+
+    public static ArrayListProductDao getInstance() {
+        return InstanceHolder.instance;
+    }
+
+    private static final class InstanceHolder {
+        static final ArrayListProductDao instance = new ArrayListProductDao();
+    }
 
-    public ArrayListProductDao() {
-        saveSampleProducts();
+    private ArrayListProductDao() {
     }
 
     @Override
-    public synchronized Product getProduct(Long id) throws RuntimeException {
+    public synchronized Product getProduct(Long id) {
+
+        if (id == null) {
+            throw new IllegalArgumentException();
+        }
+
         return products.stream()
                 .filter(product -> id.equals(product.getId()))
                 .findAny()
-                .orElseThrow(() -> new RuntimeException("Item not found"));
+                .orElseThrow(() -> new ProductNotFoundException(id));
     }
 
     @Override
-    public synchronized List<Product> findProducts() {
-        return products.stream()
-                .filter(product -> product.getPrice() != null)
-                .filter(product -> product.getStock() > 0)
-                .collect(Collectors.toList());
+    public synchronized List<Product> findProducts(String query, SortField sortField, SortOrder sortOrder) {
+        List<Product> returnProducts = findProducts(query);
+        if (sortField == null) {
+            return returnProducts;
+        }
+
+        SortOrder orderType;
```

Comment:
Since JDK 9 you have Objects.requireNonNullElse. The 6 lines can be replaced with one
`sortOrder = Objects.requireNonNullElse(sortOrder, SordOrder.ASC);`